Update .ml-config mappings before indexing job, datafeed or df analytics config

przemekwitek opened the pull request:
This PR updates .ml-config index mappings before indexing job, datafeed or df analytics config.

Without it, if the .ml-config index was created in the old version, and the first config with a new field is indexed, the default mapping (i.e. type "keyword") is applied for this new field as implemented in ElasticsearchMappings::addDefaultMapping. This default behavior is usually not desirable.

Related to https://github.com/elastic/elasticsearch/issues/44263

Commit:
Update .ml-config mappings before indexing job, datafeed or df analytics config

elasticmachine commented:
Pinging @elastic/ml-core

droberts195 commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutDataFrameAnalyticsAction.java:
```diff
@@ -150,6 +164,19 @@ private void handlePrivsResponse(String username, DataFrameAnalyticsConfig memor
         }
     }
 
+    public void updateDocMappingAndPutConfig(DataFrameAnalyticsConfig config,
```
Does this need to be `public`?

droberts195 reviewed (commented):
Thanks for doing this so quickly.

My only code comments are minor (and 3 are the same thing).

I think the really important thing is to validate this using an index created in a 6.6 cluster.  You won't be able to do this on the master branch.

One way to do it would be:

1. Create the 7.3 backport branch now and cherry-pick the commits from your master PR branch to that branch hanging off of 7.3.
2. Download a 6.6 snapshot build, for example from https://prelert-artifacts.s3-eu-west-1.amazonaws.com/maven/org/elasticsearch/ml/ml/6.6.3-SNAPSHOT/elasticsearch-6.6.3-SNAPSHOT.zip
3. Extract and run the 6.6 snapshot, and create an ML job (anything will do)
4. Shut down the ES 6.6 process
5. With the branch from step 1 checked out, run `./gradlew :distribution:archives:darwin-tar:build` to build a snapshot distribution of that branch
6. Extract the 7.3 snapshot you just built and move the `data` directory from the 6.6 snapshot install into the 7.3 install
7. Start up the 7.3 install and create an ML data frame analytics job (anything will do)
8. Check the mappings on the `.ml-config` index - have they been updated with the correct mappings for data frame analytics jobs?

In particular it's critical that `n_neighbors` is an `integer` and `feature_influence_threshold` is a `double`.  If the fix didn't work as expected they'll be `keyword`.

Once you've seen manually that the fix works it should be possible to automate it by adding a new test in `x-pack/qa/full-cluster-restart/src/test/java/org/elasticsearch/xpack/restart`.  We already have `MlMigrationFullClusterRestartIT` but you could add a new one that:

1. Creates an anomaly detector job in the old cluster
2. Creates a data frame analytics job in the new cluster
3. Asserts on the new cluster mappings after the data frame analytics job is successfully created

However, this test will only validate the 6.6 to 7.3 case when run on the 7.3 branch.  When run on the master branch it will validate a 7.x to 8.0 upgrade, which will not currently prove that the mappings updater code is working.  This is why I think it's so crucial to do a manual test with 6.6 and 7.3 before merging this PR.

przemekwitek reviewed (commented):
I applied the suggested fixes and started manual testing process.

droberts195 reviewed (approved):
LGTM assuming the 6.6->7.3 test works

przemekwitek commented:
run elasticsearch-ci/default-distro

przemekwitek commented:
run elasticsearch-ci/1 
run elasticsearch-ci/2

przemekwitek commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportPutDatafeedAction.java:
```diff
@@ -80,6 +84,7 @@ public TransportPutDatafeedAction(Settings settings, TransportService transportS
         this.datafeedConfigProvider = new DatafeedConfigProvider(client, xContentRegistry);
         this.jobConfigProvider = new JobConfigProvider(client, xContentRegistry);
         this.xContentRegistry = xContentRegistry;
+        clusterService.addListener(event -> clusterState = event.state());
```
> I think we should set the initial state as well, so `clusterState = clusterService.state()` immediately above this line. I guess in reality it doesn't matter as it could be impossible not to get some other cluster state update before a new datafeed is put. But it seems best to declare the intent that `clusterState` should not be `null`.

I had to revert this change in all the three places as the constructor is apparently too early to fetch the cluster state and many tests are failing with the exception: "initial cluster state not set yet" thrown from ClusterApplierService::state.
Instead, I skip updating mappings if clusterState is null and log warning. Please LMK if you have other idea on how to solve this problem.

droberts195 replied:
Thanks.  This sounds like the best approach.

davidkyle replied:
Why not used the `clusterstate` passed as an argument to `masterOperation()` 

`clusterState` can change while this action is executing if `clusterService` calls the listener, so you could be in a situation where one code branch it taken based on current clusterstate then the state changes and that code branch is invalid.

przemekwitek replied:
Thanks for this suggestion @davidkyle. It definitely makes sense for TransportPutJobAction and TransportPutDatafeedAction which are both subclasses of TransportMasterNodeAction.
However, I'm not sure what to do with TransportPutDataFrameAnalyticsAction which is HandledTransportAction and doesn't seem to have any handle on clusterState other than through ClusterService listener...

davidkyle replied:
You can grab the cluster state from the cluster service at any time. State is locally cached by `ClusterService` so it is a fast call. 

`ClusterState state = clusterService.state();` 

Can be called in `doExecute` or where ever needed

davidkyle replied:
Don't take the state in the constructor though as the object may be constructed long before `doExecute` is called, which I guess is why you got the _initial cluster state not set yet_ error.

przemekwitek replied:
Done.

Commit:
Do not fetch cluster state prematurely.
Log warning if it was impossible to obtain cluster state before updating mappings

Commit:
Reuse exception message's parsing logic from putJobListener in addDocMappingsListener
This fixes testCreateJob_WithClashingFieldMappingsFail tests

Commit:
Use cluster state directly when it is provided by masterOperation

Commit:
Fetch cluster state right before updating mappings

davidkyle reviewed (approved):
LGTM

przemekwitek commented:
> LGTM assuming the 6.6->7.3 test works

I've verified manually that the .ml-config has correct mappings after the upgrade.
